bundles/org.openhab.binding.linuxinput/src/main/java/org/openhab/binding/linuxinput/internal/LinuxInputHandler.java

```diff
@@ -0,0 +1,205 @@
+/**
+ * Copyright (c) 2010-2019 Contributors to the openHAB project
+ *
+ * See the NOTICE file(s) distributed with this work for additional
+ * information.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0
+ *
+ * SPDX-License-Identifier: EPL-2.0
+ */
+package org.openhab.binding.linuxinput.internal;
+
+import org.eclipse.jdt.annotation.NonNullByDefault;
+import org.eclipse.jdt.annotation.Nullable;
+import org.eclipse.smarthome.core.library.CoreItemFactory;
+import org.eclipse.smarthome.core.library.types.OpenClosedType;
+import org.eclipse.smarthome.core.library.types.StringType;
+import org.eclipse.smarthome.core.thing.*;
+import org.eclipse.smarthome.core.thing.binding.builder.ChannelBuilder;
+import org.eclipse.smarthome.core.thing.binding.builder.ThingBuilder;
+import org.eclipse.smarthome.core.types.Command;
+import org.openhab.binding.linuxinput.internal.evdev4j.EvdevDevice;
+import org.openhab.binding.linuxinput.internal.evdev4j.jnr.EvdevLibrary;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.io.IOException;
+import java.nio.channels.SelectionKey;
+import java.nio.channels.Selector;
+import java.util.*;
+
+import static org.openhab.binding.linuxinput.internal.LinuxInputBindingConstants.*;
+
+/**
+ * Handler for Linux Input devices.
+ *
+ * @author Thomas Weißschuh - Initial contribution
+ */
+@NonNullByDefault
+public final class LinuxInputHandler extends DeviceReadingHandler {
+
+    private final Logger logger = LoggerFactory.getLogger(LinuxInputHandler.class);
+
+    private @NonNullByDefault({}) Map<Integer, Channel> channels;
+    private @NonNullByDefault({}) Channel keyChannel;
+    private @NonNullByDefault({}) EvdevDevice device;
+    private final @NonNullByDefault({}) String defaultLabel;
+
+    private @NonNullByDefault({}) LinuxInputConfiguration config;
+
+    public LinuxInputHandler(Thing thing, @Nullable String defaultLabel) {
+        super(thing);
+        this.defaultLabel = defaultLabel;
+    }
+
+    @Override
+    public void handleCommand(ChannelUID channelUID, Command command) {
+        /* no commands to handle */
+    }
+
+    @Override
+    boolean immediateSetup() {
+        config = getConfigAs(LinuxInputConfiguration.class);
+        channels = Collections.synchronizedMap(new HashMap<>());
```

Comment:
If you just initialize these at the declaration or in the constructor they will be `null`-safe.

Reply:
The config and therefore the channels should be reread everytime the thing is reinitialized (fe because the config changed), no?

Reply:
Yes `config` should be kept here. But `channels` and `keyChannel` can be moved. You can clear `channels` here instead of reassigning a new map if it depends on the configuration. If the configuration has no effect on the channels no clear is needed (I didn't look deeply into how you've use the field `channels` so don't know if it needs clearing or not).